rfcs/20250430-host-side-scalars/README.md

```diff
@@ -0,0 +1,155 @@
+        dnnl_memory_desc_t *memory_desc, dnnl_data_type_t data_type);
+```
+
+Internal considerations: we would use a new `format_kind` with no
+associated `format_desc`. This would avoid aliasing host-side scalars
+with blocked descriptors, and hence host-side scalars will need to be
+explicitely supported by implementations.
+
+For users passing primitive objects around and querying them, it makes
+sense to expose this format kind, so users can query if a given input
+is a host-side scalar, and pass it accordingly to primitive execution.
+
+
+Pros:
+- avoids user from recreating primitives when scalars are changed
+- simple enabling internally
+
+Cons:
+- ties the memory descriptor to a specific device (host) which we
+  avoided so far.
+
+
+### Option 3: explicitely take engine kind in memory desc.
+
+Here, we would allow users to pass any memory on host or device.
+This would officially tie a given memory descritor to a device kind.
+
+For compatibility reasons, we could either use `undef` to specify that
+a memory desc will be ties to the same engine kind as the
+implementation, or introduce a new `engine_kind` for that purpose.
+
+A few points need to be made:
+- mixing host and device memory is already possible through usm shared
+  memory for both sycl and opencl runtime.
+- there is no clear benefit to mix host and device memory, except for
+  the case where the value to pass is a scalar, in which case it can
+  be passed to device kernel as a parameter (instead of passing a
+  pointer).
+- even if two memories share the same engine kind, it does not
+  guarentee that they will both be accessible from the same context
+  (they might be tied to different engine objects).
+
+As a result, we don't recommend to specify engine kind for memory
+descriptors as it would likely provide little benefit over host-side
+scalar specific memory descriptor, but being more general, it would
+also come with more complications.
+
+Pros:
+- clear semantic: all memory descritors are tied to an engine kind
+
+Cons:
+- full buffers sharing across devices can already be achieved through runtime mecanisms.
+- no clear benefit for performance other than specific scalar case.
+
+## How should host-side scalar be passed to execute call?
+
+Currently, the oneDNN primitive API only accepts `dnnl::memory` objects
+(through an `unordered_map` in C++ interface, or an array of
+`arg_kind`/`memory` for C interface).
+
+We have two paths:
+- encapsulate host-side scalars in memory objects: this would create
+  overhead for passing a scalar but would not require execution API
+  changes. There is a question of object lifetime to clarify here: for
+  host-side scalar, we have to guarentee that they are not accessed
+  asynchronously, otherwise users will incur extra overhead of
+  allocation and copy for a single scalar.
+- expose new execute function to allow passing scalars. For both C and
+  C++ APIs, it would require to take extra unordered_map for scalar
+  arguments.
```

Comment:
It seems to me this part should be extended further with API examples.
If the proposal is to put scalars into a map, what the execution argument is gonna be for it? A new one, or re-used? If re-used, what kind of internal changes what be required to distinguish regular memory from a scalar?
How the `dnnl::memory` object creation with host_scalar would look like? New API, or re-use one? If re-used, is engine redundant there? Or the ctor with memory allocation, which kills the whole purpose of the feature?

If the execute function allows to split constants and memories, not quite clear how the implementation should identify them which comes where if there are more than one scalar.

Reply:
hi @dzarukin, if I understood correctly, the ask is only for 2nd option, which is currently not what is proposed, but ~possible alternative? would you still like to see that level of details? do you have any concerns with recommended option (which is having scalars as memory objects)?

Reply:
hey @dzarukin as for extending primitive execution APIs, we didn't explore this option further in details, as the recommendation is to get with mem descr / mem constructor changes. Let me know if you want the other option considered instead. 
As for the memory object creation, I've put more details / how the API would look like and you could also follow the link to the POC.

Reply:
resolving as outdated